### PR TITLE
 [FEATURE][processing] New algorithm to categorize a layer using a style XML file

### PR DESCRIPTION
**Pull request by nyalldawson**
This algorithm addresses a use case which is VERY difficult to handle in QGIS at the moment -- provide a set of symbols which need to be matched to a preset set of values. This isn't always possible to do using QML files, because the field names containing the values may vary, and QML files also apply EVERYTHING, not just a renderer's settings. Additionally, QML doesn't handle the situation where a style library contains many symbols which don't apply to a single layer -- i.e. a style database designed for use across a set of layers.

So, this new algorithm sets a vector layer's renderer to a categorized renderer using matching symbols from a style database.

The specified expression (or field name) is used to create categories for the renderer. A category will be created for each unique value within the layer.

Each category is individually matched to the symbols which exist within the specified QGIS XML style database. Whenever a matching symbol name is found, the category's symbol will be set to this matched symbol.

The matching is case-insensitive by default, but can be made case-sensitive if required.

Optionally, non-alphanumeric characters in both the category value and symbol name can be ignored while performing the match. This allows for greater tolerance when matching categories to symbols.

If desired, tables can also be output containing lists of the categories which could not be matched to symbols, and symbols which were not matched to categories.

Screencast:

peek 2018-09-07 14-53

*This algorithm is best paired with a certain [thrash metal side-project](https://github.com/nyalldawson/slyr) I've been working on lately.*

### CODE REVIEW

File: src/core/symbology/qgscategorizedsymbolrenderer.cpp [hunk 1]
```diff
@@ -956,3 +957,65 @@ QgsDataDefinedSizeLegend *QgsCategorizedSymbolRenderer::dataDefinedSizeLegend()
 {
   return mDataDefinedSizeLegend.get();
 }
+
+int QgsCategorizedSymbolRenderer::matchToSymbols( QgsStyle *style, const QgsSymbol::SymbolType type, QVariantList &unmatchedCategories, QStringList &unmatchedSymbols, const bool caseSensitive, const bool useTolerantMatch )
+{
+  if ( !style )
```
Comment (nirvn) [inline src/core/symbology/qgscategorizedsymbolrenderer.cpp]: Would it be useful to rely on the user's saved styles when no QgsStyle object is provided instead? I don't have any particular scenario in mind, just throwing the idea at you.

Comment (nyalldawson) [reply src/core/symbology/qgscategorizedsymbolrenderer.cpp]: That's a good idea!

**Comment by nirvn** [comment]
Looking good

File: python/core/auto_generated/symbology/qgscategorizedsymbolrenderer.sip.in [hunk 2]
```diff
@@ -255,6 +255,26 @@ Returns configuration of appearance of legend when using data-defined size for m
...
+    int matchToSymbols( QgsStyle *style, QgsSymbol::SymbolType type,
+                        QVariantList &unmatchedCategories /Out/, QStringList &unmatchedSymbols /Out/, bool caseSensitive = true, bool useTolerantMatch = false );
+%Docstring
+Replaces category symbols with the symbols from a ``style`` that have a matching
+name and symbol ``type``.
+
+The ``unmatchedCategories`` list will be filled with all existing categories which could not be matched
+to a symbol in ``style``.
+
+The ``unmatchedSymbols`` list will be filled with all symbol names from ``style`` which were not be matched
```
Comment (DelazJ) [inline python/core/auto_generated/symbology/qgscategorizedsymbolrenderer.sip.in]: were not **be** matched?

**Pull request merged**